library/std/src/io/mod.rs

```diff
@@ -464,7 +464,8 @@ pub(crate) fn default_read_to_end<R: Read + ?Sized>(
             }
         }
 
-        if buf.len() == buf.capacity() {
+        // Ensure there's at least enough space for one UTF-8 encoded code point.
+        if buf.spare_capacity_mut().len() < char::MAX_LEN_UTF8 {
             // buf is full, need more space
             buf.try_reserve(PROBE_SIZE)?;
         }
```

Comment:
I think that this the condition could even be replaced by an unconditional call to `try_reserve(PROBE_SIZE)`.

Reply:
That might lead to some unnecessary reallocations in some edge-cases, see #89165 (also mentioned in the function's top comment). It's tricky to get these tradeoffs right.

Maybe we can limit this to cfg(windows)? Specialization could work too but feels a bit overkill here.

Reply:
Also, please add the issue to the list of masters this function is serving. Anyone who wants to touch it will have to keep track of all these things.